### PR TITLE
Remove six.Iterator (py3+) and combine six fix with class fix

### PR DESCRIPTION
**Pull request by mxr**
Part of #59.

### CODE REVIEW

File: tests/pyupgrade_test.py [hunk 1]
```diff
@@ -1092,6 +1092,10 @@ def test_fix_six(s, expected):
             'class C(object, metaclass=ABCMeta): pass',
             'class C(metaclass=ABCMeta): pass',
         ),
+        (
+            'class C(six.Iterator, metaclass=ABCMeta): pass',
```
Comment (asottile) [inline tests/pyupgrade_test.py]: should also handle `from six import Iterator`  `class C(Iterator, ...):`

I wonder what happens if

```python
class C(Iterator, object):
```

🤔

Comment (mxr) [reply tests/pyupgrade_test.py]: Good point, let me try to handle those cases

**Comment by hugovk** [comment]
@asottile I suggest updating the CI config to allow testing of all feature branches, so we know PRs pass tests.

**Comment by asottile** [comment]
> @asottile I suggest updating the CI config to allow testing of all feature branches, so we know PRs pass tests.

weird, I wonder if azure just dropped the webhook here, it *is* supposed to test PRs (in fact, *only* PRs :laughing:): 

For instance https://github.com/asottile/pyupgrade/pull/115

though maybe that's because I ran it :thinking:

**Comment by asottile** [comment]
ok I had to click a button, seems to work now: #118

**Comment by hugovk** [comment]
I think removing the branch restriction would let it build all branches automatically:

https://github.com/asottile/pyupgrade/blob/0df1eac05c9c37e2bfecf11895aee53a34c33afd/azure-pipelines.yml#L2-L3

**Comment by mxr** [comment]
@asottile let me know if the changes look good. I can't re-request a review on GitHub.

**Comment by asottile** [comment]
github actually just added that as a feature:

Screenshot from 2019-03-20 08-25-39

(the little refresh button)

File: pyupgrade.py [hunk 2]
```diff
@@ -852,26 +852,56 @@ def _fix_super(contents_text):
     return tokens_to_src(tokens)
 
 
-class FindNewStyleClasses(ast.NodeVisitor):
+class SixVisitor(ast.NodeVisitor):
```
Comment (asottile) [inline pyupgrade.py]: instead of inheritance, let's combine `SixVisitor` and `FindsClasses` (I was actually going to do that as a followup anyway).  This will also improve performance of pyupgrade by ~15% as well :tada:

File: pyupgrade.py [hunk 3]
```diff
@@ -852,26 +852,56 @@ def _fix_super(contents_text):
...
-            if isinstance(base, ast.Name) and base.id == 'object':
+            if (
+                (
+                    isinstance(base, ast.Name) and
+                    (
+                        base.id == 'object' or
+                        (
+                            base.id == 'Iterator' and
+                            'Iterator' in self.six_from_imports
+                        )
```
Comment (asottile) [inline pyupgrade.py]: let's keep these cases separate just so it's a little easier to see what's going on -- I realize you golfed down the `isinstance(..., ast.Name)` but I think it's less readable now.

Also let's do `base.id in self.six_from_imports` for the last condition

Comment (mxr) [reply pyupgrade.py]: sure

File: pyupgrade.py (hunk 2 above)
Comment (mxr) [reply pyupgrade.py]: I'll have you know this is enterprise-grade inheritance

**Comment by mxr** [comment]
I don't see  the little refresh button, only the "speech bubble" button :/

File: pyupgrade.py (hunk 2 above)
Comment (mxr) [reply pyupgrade.py]: What about FindSixUsage?

**Comment by asottile** [comment]
> I don't see the little refresh button, only the "speech bubble" button :/

riiip, maybe you're not in the experiment bucket

File: pyupgrade.py (hunk 2 above)
Comment (asottile) [reply pyupgrade.py]: yeah, combine all 3

**Comment by mxr** [comment]
@asottile combined all 3, lmk if that's what you had in mind

**Comment by mxr** [comment]
@asottile do you have an expected merge timeline?

**Comment by asottile** [comment]
> @asottile do you have an expected merge timeline?

I can't if the tests are failing :P

**Comment by mxr** [comment]
confused the green approval check with the green tests passing check 😭 and the tests (that i ran) passed on my machine

**Comment by asottile** [comment]
> confused the green approval check with the green tests passing check sob and the tests (that i ran) passed on my machine

looks like only python2.x is failing

File: pyupgrade.py [hunk 4]
```diff
@@ -1034,13 +942,23 @@ def visit_ClassDef(self, node):
             ):
                 self.remove_decorators.add(_ast_to_offset(decorator))
 
-        self.generic_visit(node)
+        for i, base in enumerate(node.bases):  # pragma: no cover (py3+)
```
Comment (asottile) [inline pyupgrade.py]: hmmmm, this seems not right -- it didn't need this comment on the left, why the change now?

Comment (mxr) [reply pyupgrade.py]: Let me scope this down

**Comment by mxr** [comment]
@asottile updated

**Comment by mxr** [comment]
did i forget to push?

**Comment by asottile** [comment]
> did i forget to push?

looks like it, yeah -- I still see the `no cover`s

**Comment by mxr** [comment]
pushed 🤦‍♂️

**Comment by asottile** [comment]
wompwomp

```python
import six


class C(object, six.Iterator): pass
```

```console
$ pyupgrade  t.py  --py3-plus
Traceback (most recent call last):
  File "/home/asottile/workspace/pyupgrade/venv/bin/pyupgrade", line 11, in <module>
    load_entry_point('pyupgrade', 'console_scripts', 'pyupgrade')()
  File "/home/asottile/workspace/pyupgrade/pyupgrade.py", line 1402, in main
    ret |= fix_file(filename, args)
  File "/home/asottile/workspace/pyupgrade/pyupgrade.py", line 1376, in fix_file
    contents_text = _fix_six_and_classes(contents_text)
  File "/home/asottile/workspace/pyupgrade/pyupgrade.py", line 1135, in _fix_six_and_classes
    while tokens[j].src != ',':
IndexError: list index out of range
```

**Comment by mxr** [comment]
something something unpaid labor

**Comment by mxr** [comment]
@asottile still haven't had a chance to look at it. Would separating the six fixer and the class fixer be an option..?

**Comment by asottile** [comment]
I don't think that helps -- I think all that needs to happen here is to make the class killer not worry about the class count and always look forward for parens or commas and then make a decision

**Comment by mxr** [comment]
I see

**Comment by mxr** [comment]
Thanks!!!
